docs/reposync.1
.\" reposync
.TH "reposync" "1" "27 April 2007" "" ""
.SH "NAME"
reposync \- synchronize yum repositories to a local directory
.SH "SYNOPSIS"
\fBreposync\fP [options]
.SH "DESCRIPTION"
\fBreposync\fP is used to synchronize a remote yum repository to a local
directory, using yum to retrieve the packages.
.SH "OPTIONS"
.IP "\fB\-h, \-\-help\fP"
Display a help message, and then quit.
.IP "\fB\-c CONFIG, \-\-config=CONFIG\fP"
Config file to use (defaults to /etc/yum.conf).
.IP "\fB\-a ARCH, \-\-arch=ARCH\fP"
Act as if running the specified arch (default: current arch, note: does
not override $releasever. x86_64 is a superset for i*86.).
.IP "\fB\-\-source\fP"
Also download .src.rpm files.
.IP "\fB\-r REPOID, \-\-repoid=REPOID\fP"
Specify repo ids to query, can be specified multiple times (default is
all enabled).
.IP "\fB\-e\ CACHEDIR, \-\-cachedir CACHEDIR\fP"
Directory in which to store metadata.
.IP "\fB\-t, \-\-tempcache\fP"
Use a temp dir for storing/accessing yum-cache.
.IP "\fB\-d, \-\-delete\fP"
Delete local packages no longer present in repository.
.IP "\fB\-p DESTDIR, \-\-download_path=DESTDIR\fP"
Path to download packages to: defaults to current directory.
.IP "\fB\-\-norepopath\fP"
Don't add the reponame to the download path.
Can only be used when syncing a single repository (default is
to add the reponame).
.IP "\fB\-g, \-\-gpgcheck\fP"
Remove packages that fail GPG signature checking after downloading.
exit status is '1' if at least one package was removed.
.IP "\fB\-u, \-\-urls\fP"
Just list urls of what would be downloaded, don't download.
.IP "\fB\-l, \-\-plugins\fP"
Enable yum plugin support.
.IP "\fB\-m, \-\-downloadcomps\fP"
Also download comps.xml.
.IP "\fB\-\-download-metadata\fP"
Download all the non-default metadata
.IP "\fB\-n, \-\-newest\-only\fP"
Download only newest packages per-repo.
.IP "\fB\-q, \-\-quiet\fP"
Output as little information as possible.
.IP "\fB\-\-allow-path-traversal\fP"
Allow packages stored outside their repo directory to be synced.
These are packages that are referenced in metadata by using absolute paths or
up-level ".." symbols, and are normally skipped by \fBreposync\fR for security
reasons.

\fBCAUTION:\fR Using this option has potential security implications since, by
providing malicious repodata, an attacker could make \fBreposync\fR write to
arbitrary locations on the file system that are accessible by the user running
it.
.SH "EXAMPLES"
.IP "Sync all packages from the 'updates' repo to the current directory:"
\fB reposync \-\-repoid=updates\fP
.IP "Sync only the newest packages from the 'updates' repo to the current directory:"
\fB reposync \-n \-\-repoid=updates\fP
.IP "Sync packages from the 'updates' and 'extras' repos to the current directory:"
\fB reposync \-\-repoid=updates \-\-repoid=extras\fP
.IP "Sync all packages from the 'updates' repo to the \fBrepos\fP directory:"
\fB reposync \-p repos \-\-repoid=updates\fP
.IP "Sync all packages from the 'updates' repo to the \fBrepos\fP directory excluding x86_64 arch. Edit \fI/etc/yum.conf\fR adding option \fBexclude=*.x86_64\fR. Then: 
\fBreposync \-p repos \-\-repoid=updates\fP
.SH "FILES"
\fBreposync\fP uses the yum libraries for retrieving information and
packages. If no configuration file is specified, the default yum
configuration will be used.
.PP
.nf 
/etc/yum.conf
/etc/yum/repos.d/
.fi
.SH "SEE ALSO"
.nf
.I yum.conf (5)
http://yum.baseurl.org/
.fi
.SH "AUTHORS"
.nf 
See the Authors file included with this program.
.fi

.PP 
.SH "BUGS"
There are of course no bugs, but should you find any, you should first
consult the FAQ section on http://yum.baseurl.org/wiki/Faq and if unsuccessful
in finding a resolution contact the mailing list: yum-devel@lists.baseurl.org.
To file a bug use http://bugzilla.redhat.com for Fedora/RHEL/Centos
related bugs and http://yum.baseurl.org/report for all other bugs.

.fi


ChangeLog
2011-08-10  Tim Lauridsen <timlau@fedoraproject.org>

	* yum-utils.spec: bumped yum-utils version to 1.1.31

2011-08-10  Tim Lauridsen <timlau@fedoraproject.org>

	* ChangeLog: updated ChangeLog

2011-08-10  Tim Lauridsen <timlau@fedoraproject.org>

	* yum-utils.spec: bumped yum-utils version to 1.1.31

2011-08-03  Shawn Starr <shawn.starr@rogers.com>

	* docs/yum-utils.1: Update yum-utils man page, to add new utils.

2011-07-25  James Antill <james@and.org>

	Merge branch 'master' of
	ssh://yum.baseurl.org/srv/projects/yum/git/yum-utils  * 'master' of
	ssh://yum.baseurl.org/srv/projects/yum/git/yum-utils: (3 commits)
	typo leaving .conf in ybc.conf.yumvar should just be ybc.yumvar
	...

2011-07-25  James Antill <james@and.org>

	* plugins/aliases/aliases: Add some new aliases, fix chkD.

2011-07-25  Seth Vidal <skvidal@fedoraproject.org>

	* yum-config-manager.py: typo leaving .conf in ybc.conf.yumvar
	should just be ybc.yumvar

2011-07-12  James Antill <james@and.org>

	* plugins/security/security.py: Also auto-apply --security when
	given --sec-severity for other commands.

2011-07-08  Ville Skyttä <ville.skytta@iki.fi>

	* yum-utils.bash: Add repoquery @group completion.

2011-07-08  Ville Skyttä <ville.skytta@iki.fi>

	* repoquery.py: Fix "nitmes" typo in previous commit.

2011-07-06  James Antill <james@and.org>

	* repoquery.py: Fix typo

2011-06-29  James Antill <james@and.org>

	* repoquery.py: Allow @foo groups as a packages selector in
	repoquery.

2011-06-25  Ville Skyttä <ville.skytta@iki.fi>

	* plugins/aliases/aliases.py, plugins/security/security.py,
	plugins/versionlock/versionlock.py: Remove some unused local
	variables and imports flagged by pyflakes.

2011-06-24  Ville Skyttä <ville.skytta@iki.fi>

	* debuginfo-install.py, package-cleanup.py, yum-builddep.py,
	yum-complete-transaction.py: Check doUtilBuildTransaction return
	value, exit early on failure.  BZ 716267

2011-06-22  James Antill <james@and.org>

	* plugins/security/security.py: Try to be clever when sorting the
	security severities.

2011-06-17  Raman Gupta <raman@rocketraman.com>

	* plugins/merge-conf/merge-conf.py: After (external) merge is
	complete, give user option to delete other file

2011-06-17  Raman Gupta <raman@rocketraman.com>

	* plugins/merge-conf/merge-conf.py: Support meld as well as vimdiff
	for merging  When using meld, the left side has the "other file"
	i.e. the file with the .rpmnew or .rpmsave extension. The right side
	has the "final file" i.e. the target file into which changes have to
	be merged.  If the rpm action was to save the existing conf file,
	then the left side will be the original conf file (.rpmsave) and the
	right side will be the new rpm-based configuration. The user should
	merge any changes they want to keep from their old .rpmsave from
	left to right.  If the rpm action was to keep the existing conf
	file, then the left side will be the new conf file (.rpmnew) and
	right side will be the current local configuration. The user should
	merge any changes in the new rpm file into their local config from
	left to right.  Ideally these operations would be 3-way diffs
	between the local file and the new rpm config, using the original
	rpm config as the common merge base. However, the original rpm
	config is not available to yum and therefore a 3-way diff is not
	possible. Perhaps a future enhancement would be to use a .orig file,
	if present, as the merge base for meld.  When using vimdiff, the
	order has not changed. The first parameter to vimdiff is the "final:
	file i.e. the target file, and the second parameter is the "other"
	file i.e. the .rpmnew or .rpmsave file.

2011-06-10  James Antill <james@and.org>

	* repodiff.py: Huge speedup for large number of add/remove obsoletes
	processing.

2011-06-09  Nick Jacek <njacek@redhat.com>

	* yumdownloader.py: Stops yumdownloader from attempting to download
	the same package twice if it is available in multiple repos. BZ
	711767

2011-06-10  James Antill <james@and.org>

	* docs/repodiff.1, repodiff.py: Add --compare-arch for the other
	side of the multilib. diff. BZ 710579.

2011-06-09  Nick Jacek <njacek@redhat.com>

	* repodiff.py: Changed archlist set up to handle multilib. BZ 710579


2011-06-01  James Antill <james@and.org>

	* docs/repoquery.1: Add releasever and setopt docs. from yum to
	repoquery. BZ 638542.

2011-05-20  Seth Vidal <skvidal@fedoraproject.org>

	* yum-utils.spec: frelling pyyaml pkg named with weird case

2011-05-20  Seth Vidal <skvidal@fedoraproject.org>

	* plugins/puppetverify/puppetverify.conf,
	plugins/puppetverify/puppetverify.py, yum-utils.spec: add
	puppetverify plugin

2011-05-18  Frank Weltersbach <frankwe@live.com>

	* docs/package-cleanup.1: Fix typo in package-cleanup man page. BZ
	705617.

2011-05-17  Seth Vidal <skvidal@fedoraproject.org>

	* repoquery.py:     - fix rh bug: #705179     - fix fmt_queryformat not being happy
	    with new 'output' formats passed       everywhere else.

2011-04-26  James Antill <james@and.org>

	* yumdownloader.py: Allow files/provides to be used in
	yumdownloader.

2011-04-26  James Antill <james@and.org>

	* yumdownloader.py: Exit with code 2, if we can't download the
	packages asked for. BZ 699469.

2011-04-26  James Antill <james@and.org>

	* plugins/auto-update-debuginfo/auto-update-debuginfo.py: try/except
	the open, for the rpmdbv cached data (a-u-d plugin).

2011-04-24  Ville Skyttä <ville.skytta@iki.fi>

	* yum-utils.bash: Add yum-debug-dump completion.

2011-04-24  Ville Skyttä <ville.skytta@iki.fi>

	* yum-debug-dump.py: Mention filename arg in yum-debug-dump usage
	message.

2011-04-24  Ville Skyttä <ville.skytta@iki.fi>

	* yum-debug-dump.py: Make yum-debug-dump.py executable for running
	in-place.

2011-04-24  Ville Skyttä <ville.skytta@iki.fi>

	* yum-utils.bash: Use functions from recent bash-completion if
	available.

2011-04-06  James Antill <james@and.org>

	* repoquery.py, yum-groups-manager.py: Convert

reposync.py
#!/usr/bin/python -tt

# This program is free software; you can redistribute it and/or modify
# it under the terms of the GNU General Public License as published by
# the Free Software Foundation; either version 2 of the License, or
# (at your option) any later version.
#
# This program is distributed in the hope that it will be useful,
# but WITHOUT ANY WARRANTY; without even the implied warranty of
# MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
# GNU General Public License for more details.
#
# You should have received a copy of the GNU General Public License
# along with this program; if not, write to the Free Software Foundation,
# Inc., 51 Franklin Street, Fifth Floor, Boston, MA 02110-1301 USA.
# copyright 2006 Duke University
# author seth vidal

# sync all or the newest packages from a repo to the local path
# TODO:
#     have it print out list of changes
#     make it work with mirrorlists (silly, really)
#     man page/more useful docs
#     deal nicely with a package changing but not changing names (ie: replacement)

# criteria
# if a package is not the same and smaller then reget it
# if a package is not the same and larger, delete it and get it again
# always replace metadata files if they're not the same.





import os
import sys
import shutil
import stat

from optparse import OptionParser
from urlparse import urljoin

from yumutils.i18n import _

import yum
import yum.Errors
from yum.packageSack import ListPackageSack
import rpmUtils.arch
import logging
from urlgrabber.progress import TextMeter, TextMultiFileMeter
import urlgrabber

class RepoSync(yum.YumBase):
    def __init__(self, opts):
        yum.YumBase.__init__(self)
        self.logger = logging.getLogger('yum.verbose.reposync')
        self.opts = opts

def localpkgs(directory):
    names = os.listdir(directory)

    cache = {}
    for name in names:
        fn = os.path.join(directory, name)
        try:
            st = os.lstat(fn)
        except os.error:
            continue
        if stat.S_ISDIR(st.st_mode):
            subcache = localpkgs(fn)
            for pkg in subcache.keys():
                cache[pkg] = subcache[pkg]
        elif stat.S_ISREG(st.st_mode) and name.endswith(".rpm"):
            cache[name] = {'path': fn, 'size': st.st_size, 'device': st.st_dev}
    return cache

def is_subpath(path, root):
    root = os.path.realpath(root)
    path = os.path.realpath(os.path.join(root, path))
    # join() is used below to ensure root ends with a slash
    return path.startswith(os.path.join(root, ''))

def parseArgs():
    usage = _("""
    Reposync is used to synchronize a remote yum repository to a local 
    directory using yum to retrieve the packages.
    
    %s [options]
    """) % sys.argv[0]

    parser = OptionParser(usage=usage)
    parser.add_option("-c", "--config", default='/etc/yum.conf',
        help=_('config file to use (defaults to /etc/yum.conf)'))
    parser.add_option("-a", "--arch", default=None,
        help=_('act as if running the specified arch (default: current arch, note: does not override $releasever. x86_64 is a superset for i*86.)'))
    parser.add_option("--source", default=False, dest="source", action="store_true",
                      help=_('operate on source packages'))
    parser.add_option("-r", "--repoid", default=[], action='append',
        help=_("specify repo ids to query, can be specified multiple times (default is all enabled)"))
    parser.add_option("-e", "--cachedir",
        help=_("directory in which to store metadata"))
    parser.add_option("-t", "--tempcache", default=False, action="store_true",
        help=_("Use a temp dir for storing/accessing yum-cache"))
    parser.add_option("-d", "--delete", default=False, action="store_true",
        help=_("delete local packages no longer present in repository"))
    parser.add_option("-p", "--download_path", dest='destdir',
        default=os.getcwd(), help=_("Path to download packages to: defaults to current dir"))
    parser.add_option("--norepopath", dest='norepopath', default=False, action="store_true",
        help=_("Don't add the reponame to the download path. Can only be used when syncing a single repository (default is to add the reponame)"))
    parser.add_option("-g", "--gpgcheck", default=False, action="store_true",
        help=_("Remove packages that fail GPG signature checking after downloading"))
    parser.add_option("-u", "--urls", default=False, action="store_true",
        help=_("Just list urls of what would be downloaded, don't download"))
    parser.add_option("-n", "--newest-only", dest='newest', default=False, action="store_true",
        help=_("Download only newest packages per-repo"))
    parser.add_option("-q", "--quiet", default=False, action="store_true",
        help=_("Output as little as possible"))
    parser.add_option("-l", "--plugins", default=False, action="store_true",
        help=_("enable yum plugin support"))
    parser.add_option("-m", "--downloadcomps", default=False, action="store_true",
        help=_("also download comps.xml"))
    parser.add_option("", "--download-metadata", dest="downloadmd",
        default=False, action="store_true",
        help=_("download all the non-default metadata"))
    parser.add_option("", "--allow-path-traversal", default=False,
        action="store_true",
        help=_("Allow packages stored outside their repo directory to be synced "
               "(UNSAFE, USE WITH CAUTION!)"))
    (opts, args) = parser.parse_args()
    return (opts, args)


def main():
    (opts, dummy) = parseArgs()

    if not os.path.exists(opts.destdir) and not opts.urls:
        try:
            os.makedirs(opts.destdir)
        except OSError, e:
            print >> sys.stderr, _("Error: Cannot create destination dir %s") % opts.destdir
            sys.exit(1)

    if not os.access(opts.destdir, os.W_OK) and not opts.urls:
        print >> sys.stderr, _("Error: Cannot write to  destination dir %s") % opts.destdir
        sys.exit(1)

    my = RepoSync(opts=opts)
    my.doConfigSetup(fn=opts.config, init_plugins=opts.plugins)

    # Force unprivileged users to have a private temporary cachedir
    # if they've not given an explicit cachedir
    if os.getuid() != 0 and not opts.cachedir:
        opts.tempcache = True

    if opts.tempcache:
        if not my.setCacheDir(force=True, reuse=False):
            print >> sys.stderr, _("Error: Could not make cachedir, exiting")
            sys.exit(50)
        my.conf.uid = 1 # force locking of user cache
    elif opts.cachedir:
        my.repos.setCacheDir(opts.cachedir)

    # Lock if they've not given an explicit cachedir
    if not opts.cachedir:
        try:
            my.doLock()
        except yum.Errors.LockError, e:
            print >> sys.stderr, _("Error: %s") % e
            sys.exit(50)

    #  Use progress bar display when downloading repo metadata
    # and package files ... needs to be setup before .repos (ie. RHN/etc.).
    if not opts.quiet:
        my.repos.setProgressBar(TextMeter(fo=sys.stdout), TextMultiFileMeter(fo=sys.stdout))
    my.doRepoSetup()

    if len(opts.repoid) > 0:
        myrepos = []

        # find the ones we want
        for glob in opts.repoid:
            add_repos = my.repos.findRepos(glob)
            if not add_repos:
                print >> sys.stderr, _("Warning: cannot find repository %s") % glob
                continue
            myrepos.extend(add_repos)

        if not myrepos:
            print >> sys.stderr, _("No repositories found")
            sys.exit(1)

        # disable them all
        for repo in my.repos.repos.values():
            repo.disable()

        # enable the ones we like
        for repo in myrepos:
            repo.enable()

    # --norepopath can only be sensibly used with a single repository:
    if len(my.repos.listEnabled()) > 1 and opts.norepopath:
        print >> sys.stderr, _("Error: Can't use --norepopath with multiple repositories")
        sys.exit(1)

    try:
        arches = rpmUtils.arch.getArchList(opts.arch)
        if opts.source:
            arches += ['src']
        my.doSackSetup(arches)
    except yum.Errors.RepoError, e:
        print >> sys.stderr, _("Error setting up repositories: %s") % e
        # maybe this shouldn't be entirely fatal
        sys.exit(1)

    exit_code = 0
    for repo in my.repos.listEnabled():
        reposack = ListPackageSack(my.pkgSack.returnPackages(repoid=repo.id))

        if opts.newest:
            download_list = reposack.returnNewestByNameArch()
        else:
            download_list = list(reposack)

        if opts.norepopath:
            local_repo_path = opts.destdir
        else:
            local_repo_path = opts.destdir + '/' + repo.id

        # Ensure we don't traverse out of local_repo_path by dropping any
        # packages whose remote_path is absolute or contains up-level
        # references (unless explicitly allowed).
        # See RHBZ#1600221 for details.
        if not opts.allow_path_traversal:
            newlist = []
            skipped = False
            for pkg in download_list:
                if is_subpath(pkg.remote_path, local_repo_path):
                    newlist.append(pkg)
                    continue
                my.logger.warning(
                    _('WARNING: skipping package %s: remote path "%s" not '
                      'within repodir, unsafe to mirror locally')
                    % (pkg, pkg.remote_path)
                )
                skipped = True
            if skipped:
                my.logger.info(
                    _('You can enable unsafe remote paths by using '
                      '--allow-path-traversal (see reposync(1) for details)')
                )
            download_list = newlist

        if opts.delete and os.path.exists(local_repo_path):
            current_pkgs = localpkgs(local_repo_path)

            download_set = {}
            for pkg in download_list:
                rpmname = os.path.basename(pkg.remote_path)
                download_set[rpmname] = 1

            for pkg in current_pkgs:
                if pkg in download_set:
                    continue

                if not opts.quiet:
                    my.logger.info("Removing obsolete %s", pkg)
                os.unlink(current_pkgs[pkg]['path'])

        if opts.downloadcomps or opts.downloadmd:

            if not os.path.exists(local_repo_path):
                try:
                    os.makedirs(local_repo_path)
                except IOError, e:
                    my.logger.error("Could not make repo subdir: %s" % e)
                    my.closeRpmDB()
                    sys.exit(1)

            if opts.downloadcomps:
                wanted_types = ['group']

            if opts.downloadmd:
                wanted_types = repo.repoXML.fileTypes()

            for ftype in repo.repoXML.fileTypes():
                if ftype in ['primary', 'primary_db', 'filelists',
                             'filelists_db', 'other', 'other_db']:
                    continue
                if ftype not in wanted_types:
                    continue

                try:
                    resultfile = repo.retrieveMD(ftype)
                    basename = os.path.basename(resultfile)
                    if ftype == 'group' and opts.downloadcomps: # for compat with how --downloadcomps saved the comps file always as comps.xml
                        basename = 'comps.xml'
                    shutil.copyfile(resultfile, "%s/%s" % (local_repo_path, basename))
                except yum.Errors.RepoMDError, e:
                    if not opts.quiet:
                        my.logger.error("Unable to fetch metadata: %s" % e)

        remote_size = 0
        if not opts.urls:
            for pkg in download_list:
                local = os.path.join(local_repo_path, pkg.remote_path)
                sz = int(pkg.returnSimple('packagesize'))
                if os.path.exists(local) and os.path.getsize(local) == sz:
                    continue
                remote_size += sz

        if hasattr(urlgrabber.progress, 'text_meter_total_size'):
            urlgrabber.progress.text_meter_total_size(remote_size)

        download_list.sort(key=lambda pkg: pkg.name)
        if opts.urls:
            for pkg in download_list:
                local = os.path.join(local_repo_path, pkg.remote_path)
                if not (os.path.exists(local) and my.verifyPkg(local, pkg, False)):
                    print urljoin(pkg.repo.urls[0], pkg.remote_path)
            continue

        # create dest dir
        if not os.path.exists(local_repo_path):
            os.makedirs(local_repo_path)

        # set localpaths
        for pkg in download_list:
            pkg.localpath = os.path.join(local_repo_path, pkg.remote_path)
            pkg.repo.copy_local = True
            pkg.repo.cache = 0
            localdir = os.path.dirname(pkg.localpath)
            if not os.path.exists(localdir):
                os.makedirs(localdir)

        # use downloader from YumBase
        probs = my.downloadPkgs(download_list)
        if probs:
            exit_code = 1
            for key in probs:
                for error in probs[key]:
                    my.logger.error('%s: %s', key, error)

        if opts.gpgcheck:
            for pkg in download_list:
                result, error = my.sigCheckPkg(pkg)
                if result != 0:
                    rpmfn = os.path.basename(pkg.remote_path)
                    if result == 1:
                        my.logger.warning('Removing %s, due to missing GPG key.' % rpmfn)
                    elif result == 2:
                        my.logger.warning('Removing %s due to failed signature check.' % rpmfn)
                    else:
                        my.logger.warning('Removing %s due to failed signature check: %s' % rpmfn)
                    os.unlink(pkg.localpath)
                    exit_code = 1
                    continue

    my.closeRpmDB()
    sys.exit(exit_code)

if __name__ == "__main__":
    main()
